lib/Conversion/TorchToLinalg/Random.cpp

```diff
@@ -192,6 +199,301 @@ class ConvertAtenUniformOp : public OpConversionPattern<AtenUniformOp> {
+      Value max = b.create<arith::ConstantOp>(loc, f64Ty,
+                                              rewriter.getF64FloatAttr(1.0));
+
+      // iterate and sample class indices
+      Value result = args[0];
+      Value finalResult =
+          rewriter
+              .create<scf::ForOp>(
+                  loc, cstZero, numSamples, cstOne, ValueRange{result},
+                  [&](OpBuilder &b, Location loc, Value i, ValueRange args) {
+                    // Sample random float
+                    Value uniformSample =
+                        randomUniformF64(b, loc, i, key, min, max);
+
+                    // binary search in cdf to find our sample
+                    Value left = b.create<arith::ConstantOp>(
+                        loc, i64Ty, b.getI64IntegerAttr(0));
+                    Value right = numCategories;
+
+                    auto checkCondition = [&](OpBuilder &b, Location loc,
+                                              ValueRange vals) {
+                      Value left = vals[0];
+                      Value right = vals[1];
+
+                      // while (right > left)
+                      auto comparisonPredicate = arith::CmpIPredicateAttr::get(
+                          b.getContext(), arith::CmpIPredicate::sgt);
+                      Value loopCondition = b.create<arith::CmpIOp>(
+                          loc, comparisonPredicate, right, left);
+                      b.create<scf::ConditionOp>(loc, loopCondition, vals);
+                    };
+
+                    ValueRange whileResults =
+                        b.create<scf::WhileOp>(
+                             loc, TypeRange{i64Ty, i64Ty},
+                             ValueRange{left, right}, checkCondition,
+                             [&](OpBuilder &b, Location loc, ValueRange vals) {
+                               Value left = vals[0];
+                               Value right = vals[1];
+
+                               Value two = b.create<arith::ConstantOp>(
+                                   loc, i64Ty, b.getI64IntegerAttr(2));
+                               Value diff =
+                                   b.create<arith::SubIOp>(loc, right, left);
+                               Value diffMid =
+                                   b.create<arith::DivSIOp>(loc, diff, two);
+                               Value midPointer =
+                                   b.create<arith::AddIOp>(loc, left, diffMid);
+                               Type indexTy = b.getIndexType();
+                               Value midIndex = b.create<arith::IndexCastOp>(
+                                   loc, indexTy, midPointer);
+
+                               // branch and update search indices
+                               auto thenBlock = [&](OpBuilder &b,
+                                                    Location loc) {
+                                 // left += 1
```

Comment:
Why did you opt to only increment the left endpoint by one in this case?

Reply:
Whoops, this is a logic error - I should've done newLeft = mid_pointer + 1